### DISCUSSION TITLE
Async access to web storage in FireMonkey

### DISCUSSION THREAD

**Discussion by LucidicShadow** [erosman/support#324] [Q&A] [answered]
Hi,

Can anyone provide some clarification about what storage is available inside a FireMonkey script?
I read through the FireMonkey help and found the section about Asynchronous & Synchronous Storage, however the implications of this are unclear to me.

I'm currently working on a user script, in which I'm trying to build and access an IndexedDb, but have been facing persistent issues with database objects being undefined.
I've tried both using a wrapper library called Dexie, and writing all the database access in vanilla JS, neither seem to work.
Objects are being created and used in async methods, so functions called on them should be waiting until the promise is fulfilled, however they are clearly not waiting, as evidenced by the undefined object errors.
Oddly, I can see the database appear in the browser storage, but object stores are not being created.

Edit: I've rewritten my script using [Jake Archibald's idb library](https://github.com/jakearchibald/idb) that wraps IndexedDB in promises, using the version that creates an idb global object with the functions as methods. This appears to have solved the problem.

**Comment by erosman** [comment, marked as answer]
### GM Storage API

In legacy Firefox, storage access was synchronous. The storage API  in User-script managers were also synchronous e,g. `GM_getValue`, `GM_setValue`, `GM_listValues` & `GM_deleteValue`

Since Firefox 57, all storage operations are asynchronous for Firefox and extensions. The storage API  in User-script managers had to change for the asynchronous access.

GreaseMonkey adopted the new asynchronous API `GM.getValue`, `GM.setValue`, `GM.listValues` & `GM.deleteValue`.

TamperMonkey & ViolentMonkey decided to keep the `GM_getValue`, `GM_setValue`, `GM_listValues` & `GM_deleteValue` and create a quasi-synchronous storage access by delaying the script injection and getting the storage first and then injecting the script with its storage into the page. Although the `GM_getValue` may appear synchronous, in reality it is not. However, that approach meant that many userscript developers can continue without having to update their userscripts. 

All above managers (GM | TM | VM) manually inject userscript every time a tab navigates. FireMonkey uses the dedicated secure userScript API and passes the script injection to Firefox's native process.

Originally, FireMonkey also adapted the asynchronous only storage API, similar to GreaseMonkey, but after requests a work-around was added.

The work-around means, in synchronous `GM_getValue` & `GM_listValues`
- if userscript needs to gets the storage immediately as soon as it is injected, then it gets the data at the time of Firefox start, or userscript enabled
- if userscript needs to gets the storage later e,g, after a user click etc, then it gets the more up-to-date data at the time of tab loading

The asynchronous `GM.getValue` & `GM.setValue` in FireMonkey will always get the most up-to-date data at that moment.

### IndexedDB API

If you are using IndexedDB, it is not handled by FireMonkey. IndexedDB is an asynchronous Web API. It is also domain linked so it will only be available to the domain that it was set in. For example if a userscript has `@match` for 3 domains, if it sets IndexedDB in the first one, it will not be available in the 2nd or 3rd domains.

> [IndexedDB API](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API)
> 
> **Note:** Like most web storage solutions, IndexedDB follows a [same-origin policy](https://www.w3.org/Security/wiki/Same_Origin_Policy). So while you can access stored data within a domain, you cannot access data across different domains.

More Info: [Using IndexedDB](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API/Using_IndexedDB)

I hope that helps. Let me know if there are more questions.
